File: src/main/java/com/possible_triangle/webcommands/mixin/MinecraftServerMixin.java
package com.possible_triangle.webcommands.mixin;

import com.possible_triangle.webcommands.RestClient;
import com.possible_triangle.webcommands.SocketServer;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.Text;
import org.jetbrains.annotations.Nullable;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

import java.io.IOException;
import java.util.UUID;

@Mixin(MinecraftServer.class)
public class MinecraftServerMixin {

    @Nullable
    private SocketServer socket = null;

    @Inject(at = @At("RETURN"), method = "stop(Z)V")
    public void stop(boolean bl, CallbackInfo callback) {
        RestClient.INSTANCE.onStop((MinecraftServer) (Object) this);
        if(socket != null) {
            try {
                socket.stop();
            } catch (IOException | InterruptedException e) {
                System.err.println("An error occurred on closing the websocket");
            }
        }
    }

    @Inject(at = @At("RETURN"), method = "<init>")
    public void start(CallbackInfo callback) {
        RestClient.INSTANCE.onStart((MinecraftServer) (Object) this);
        socket = new SocketServer((MinecraftServer) (Object) this);
        socket.start();
    }

    @Inject(at = @At("RETURN"), method = "sendSystemMessage(Lnet/minecraft/text/Text;Ljava/util/UUID;)V")
    public void message(Text message, UUID sender, CallbackInfo callback) {
        if(socket != null) {
            ServerPlayerEntity player = ((MinecraftServer) (Object) this).getPlayerManager().getPlayer(sender);
            socket.sendMessage(message, player);
        }
    }

}

File: src/main/kotlin/com/possible_triangle/webcommands/mixin/MinecraftServerMixin.kt
package com.possible_triangle.webcommands.mixin

import com.possible_triangle.webcommands.SocketServer
import net.minecraft.server.MinecraftServer
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import java.util.*
import net.minecraft.text.Text

@Suppress("CAST_NEVER_SUCCEEDS")
@Mixin(MinecraftServer::class)
class MinecraftServerMixin {

    private var socket: SocketServer? = null

    @Inject(at = [At("RETURN")], method = ["Lnet/minecraft/server/MinecraftServer;stop(Z)V"])
    fun stop(bl: Boolean, callback: CallbackInfo) {
        socket?.stop()
    }

    @Inject(at = [At("RETURN")], method = ["<init>"])
    fun start(callback: CallbackInfo) {
        socket = SocketServer(this as MinecraftServer)
        socket?.start()
    }

    @Inject(at = [At("RETURN")], method = ["Lnet/minecraft/server/MinecraftServer;sendSystemMessage(Lnet/minecraft/text/Text;Ljava/util/UUID;)V"])
    fun message(message: Text, sender: UUID, callback: CallbackInfo) {
        val server = this as MinecraftServer
        val player = server.playerManager.getPlayer(sender)
        socket?.sendMessage(message, player)
    }

}